Modules/_sqlite/connection.c

```diff
@@ -561,15 +560,19 @@ _pysqlite_build_py_params(sqlite3_context *context, int argc,
             case SQLITE_FLOAT:
                 cur_py_value = PyFloat_FromDouble(sqlite3_value_double(cur_value));
                 break;
-            case SQLITE_TEXT:
-                val_str = (const char*)sqlite3_value_text(cur_value);
-                cur_py_value = PyUnicode_FromString(val_str);
-                /* TODO: have a way to show errors here */
-                if (!cur_py_value) {
-                    PyErr_Clear();
-                    cur_py_value = Py_NewRef(Py_None);
+            case SQLITE_TEXT: {
+                sqlite3 *db = sqlite3_context_db_handle(context);
+                const char *text = (const char *)sqlite3_value_text(cur_value);
+
+                if (text == NULL && sqlite3_errcode(db) == SQLITE_NOMEM) {
+                    PyErr_NoMemory();
+                    goto error;
                 }
+
+                Py_ssize_t size = sqlite3_value_bytes(cur_value);
```

Comment:
Why `sqlite3_value_bytes`? Should't you just use `val_str` and `PyUnicode_FromString`?

Reply:
It's a micro optimisation:
SQLite already looped over the input string during conversion. At this point, sqlite3_value_bytes gives us the lenght without calling strlen. No need for PyUnicode_FromString to loop over it again :)

Reply:
Also it works with strings containing NUL. @erlend-aasland, would you mind to add a test for string containing NUL?